nautilus_core/common/src/throttler/mod.rs

```diff
@@ -0,0 +1,457 @@
+            warm: false,
+            interval,
+            clock,
+            timer_name,
+            output_send,
+            output_drop,
+            callback: None,
+        }
+    }
+
+    /// Returns the number of messages in the buffer.
+    // #[must_use]
+    pub fn qsize(&self) -> usize {
+        self.buffer.len()
+    }
+
+    /// Resets the throttler.
+    pub fn reset(&mut self) {
+        self.buffer.clear();
+        self.warm = false;
+        self.recv_count = 0;
+        self.sent_count = 0;
+        self.is_limiting = false;
+    }
+
+    /// Returns the utilization of the throttler.
+    // #[must_use]
+    pub fn used(&self) -> f64 {
+        if !self.warm && self.sent_count < 2 {
+            return 0.0;
+        }
+
+        let mut diff =
+            self.clock.timestamp_ns().as_i64() - self.timestamps.back().unwrap().as_i64();
+        diff = max(0, self.interval as i64 - diff);
+        let mut used = diff as f64 / self.interval as f64;
+
+        if !self.warm {
+            used *= self.sent_count as f64 / self.limit as f64;
+        }
+
+        used
+    }
+
+    /// Sends a message.
+    pub fn send(&mut self, msg: T) {
+        self.recv_count += 1;
+
+        if self.is_limiting {
+            self.limit_msg(msg);
+            return;
+        }
+
+        let delta_next = self.delta_next();
+        if delta_next <= 0 {
+            self.send_msg(msg);
+        } else {
+            self.limit_msg(msg);
+        }
+    }
+
+    /// Returns the time until the next message can be sent.
+    // #[must_use]
+    pub fn delta_next(&mut self) -> i64 {
+        if !self.warm {
+            if self.sent_count < self.limit {
+                return 0;
+            }
+            self.warm = true;
+        }
+
+        let diff = self.clock.timestamp_ns().as_u64()
+            - self
+                .timestamps
+                .back()
+                .unwrap_or_else(|| panic!("Failed to get timestamp"))
+                .as_u64();
+
+        self.interval as i64 - diff as i64
+    }
+
+    /// Limits the message rate.
+    fn limit_msg(&mut self, msg: T) {
+        if self.output_drop.is_none() {
+            self.buffer.push_front(msg);
+            self.set_callback(|throttler, event| throttler.process(event));
+            log::debug!("Buffering {}", self.buffer.len());
+        } else {
+            log::debug!("Dropping");
+            (self.output_send)(msg);
+            self.set_callback(|throttler, event| throttler.resume(event));
+        }
+
+        if !self.is_limiting {
+            log::debug!("Limiting");
+            self.set_timer();
+            self.is_limiting = true;
+        }
+    }
+
+    fn set_timer(&mut self) {
+        if self.clock.timer_names().contains(&self.timer_name.as_str()) {
+            self.clock.cancel_timer(&self.timer_name);
+        }
+
+        let delta_next = self.delta_next();
+
+        // let callback = SafeTimeEventCallback {
```

Comment:
This PR describes the new TimeEventCallback style for clocks - https://github.com/nautechsystems/nautilus_trader/pull/1922#issue-2513546817. It will replace the `SafeTimeEventCallback` style used here.